clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1957,58 +1961,166 @@ namespace {
-          RHS->getType()->isCheckedPointerPtrType()) {
-        // ptr<T> to ptr<T> assignment, no obligation to infer any bounds for either side
-      }
-      else if (LHSType->isCheckedPointerType() ||
-               LHSType->isIntegerType() ||
-               IsBoundsSafeInterfaceAssignment(LHSType, RHS)) {
-        // Check that the value being assigned has bounds if the
-        // target of the LHS lvalue has bounds.
-        LHSTargetBounds = InferLValueTargetBounds(LHS, CSS);
-        if (!LHSTargetBounds->isUnknown()) {
-          if (E->isCompoundAssignmentOp())
-            RHSBounds = InferRValueBounds(E, CSS);
-          else
-            RHSBounds = InferRValueBounds(RHS, CSS);
+      // Recursively compute rvalue bounds for the subexpressions,
+      // without performing bounds checking.  Since TraverseStmt still checks
+      // all substatements, enabling bounds checking here could cause
+      // duplicate side effects for an expression.  In future refactoring
+      // stages, these calls to RValueBounds will be replaced with calls to
+      // TraverseStmt with side effects enabled (once TraverseStmt no longer
+      // always checks substatements).
+      BoundsExpr *LHSBounds = RValueBounds(LHS, CSS, Facts,
+                                           SideEffects::Disabled);
+      BoundsExpr *RHSBounds = RValueBounds(RHS, CSS, Facts,
+                                           SideEffects::Disabled);
 
-          if (RHSBounds->isUnknown()) {
-             S.Diag(RHS->getBeginLoc(),
-                    diag::err_expected_bounds_for_assignment)
-                    << RHS->getSourceRange();
-             RHSBounds = S.CreateInvalidBoundsExpr();
+      // Perform checking with side effects, if enabled.
+      if (SE == SideEffects::Enabled) {
+        if (E->isAssignmentOp()) {
+          QualType LHSType = LHS->getType();
+          // Bounds of the target of the lvalue
+          BoundsExpr *LHSTargetBounds = nullptr;
+          // Bounds of the right-hand side of the assignment
+          BoundsExpr *RightBounds = nullptr;
+
+          if (!E->isCompoundAssignmentOp() &&
+              LHSType->isCheckedPointerPtrType() &&
+              RHS->getType()->isCheckedPointerPtrType()) {
+            // ptr<T> to ptr<T> assignment, no obligation to infer any bounds for either side
+          }
+          else if (LHSType->isCheckedPointerType() ||
+                   LHSType->isIntegerType() ||
+                   IsBoundsSafeInterfaceAssignment(LHSType, RHS)) {
+            // Check that the value being assigned has bounds if the
+            // target of the LHS lvalue has bounds.
+            LHSTargetBounds = InferLValueTargetBounds(LHS, CSS);
+            if (!LHSTargetBounds->isUnknown()) {
+              if (E->isCompoundAssignmentOp())
+                RightBounds = InferRValueBounds(E, CSS, Facts);
+              else
+                RightBounds = InferRValueBounds(RHS, CSS, Facts);
```

Comment:
InferRValueBounds calls RValueBounds and then does an additional check.    Instead of calling InferRValueBounds, could you use the previously computed RValueBounds of RHS and do the check here?